File: day27/string_compression.py
def compress_string(s: str) -> str:
    if not s:
        print("Original: ''")
        print("Compressed: ''")
        return s

    compressed = []
    count = 1

    for i in range(1, len(s)):
        if s[i] == s[i - 1]:
            count += 1
        else:
            compressed.append(s[i - 1] + str(count))
            count = 1

    compressed.append(s[-1] + str(count))  # append last group
    result = ''.join(compressed)

    print(f"Original: {s}")
    print(f"Compressed: {result if len(result) < len(s) else s}")

    return result if len(result) < len(s) else s

# Example test
compress_string("aabcccccaaa")
compress_string("abc")

File: day27/stringCompression.cpp
#include <iostream>
#include <string>

std::string compressString(const std::string& s) {
    if (s.empty()) {
        std::cout << "Original: \"\"\nCompressed: \"\"\n";
        return s;
    }

    std::string compressed;
    int count = 1;

    for (size_t i = 1; i < s.length(); ++i) {
        if (s[i] == s[i - 1]) {
            ++count;
        } else {
            compressed += s[i - 1] + std::to_string(count);
            count = 1;
        }
    }
    compressed += s.back() + std::to_string(count);

    std::cout << "Original: " << s << "\n";
    std::cout << "Compressed: " << (compressed.length() < s.length() ? compressed : s) << "\n";

    return compressed.length() < s.length() ? compressed : s;
}

int main() {
    compressString("aabcccccaaa");
    compressString("abc");
    return 0;
}
